### ISSUE TITLE
Server-side TLS Server Name Indication (SNI) support

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4887]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4887|
|Type|enhancement|
|Created|2011-02-16 22:35:37Z|

This depends on [PyOpenSSL supporting SNI](https://bugs.launchpad.net/pyopenssl/+bug/705683).

This should modify the SSL endpoint string description syntax so that existing applications will inherit it seamlessly.

<details><summary>Searchable metadata</summary>

```
trac-id__4887 4887
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1297895737000000 1297895737000000
changetime__1560836566327149 1560836566327149
version__None None
owner__None None
cc__chris@... cc__hynek cc__mithrandi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

This ticket should be for server-side support.  I don't know what's required client-side yet.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

pyOpenSSL 0.13 supports SNI.  For a server, what you do is call `set_tlsext_servername_callback` on the `Context` instance you're using.  You pass it a callback function that will be called with a `Connection` instance.  In the callback function, you use `Connection.get_servername` to see who the client wanted to talk to.  Then, if appropriate, you construct a **new** `Context` configured however you like (including using a certificate/key which matches the client's expectations) and make the server use it using `Connection.set_context`.

While not the prettiest API (not least because it forces a lot of OpenSSL specifics onto you), it's actually all accessible already, without any particular additional support from Twisted.  So from that point of view, this ticket is actually resolved.

The client-side support is actually harder, because there currently is no way in the Twisted SSL client-side APIs to specify which hostname you want to talk to.

So with pyOpenSSL 0.13 and current Twisted, server-side is ugly but possible and client-side is not possible at all.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_3):

Thanks for this update, and thanks even more for actually implementing the requisite SNI wrappers.

I think that we should probably have some explicit facility to do SNI so that it can get communicated down to layers like HTTP without OpenSSL specifics.

Also, anyone who takes on the client portion of this implementation should be aware of [#5190](https://github.com/twisted/twisted/issues/5190).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez set owner to @glyph|
|-|-|

Assigning this to Glyph while I’m still his boss.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_10"></a>|@hynek commented|
|-|-|

Any plans to actually do this or are we going to leave txsni for a while and see how it evolves?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Oh yeah we should probably actually link to [txsni](https://warehouse.python.org/project/TxSNI/) here, shouldn't we.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_10):
> Any plans to actually do this or are we going to leave txsni for a while and see how it evolves?

I'd like to have SNI support within Twisted proper, but txsni is an extremely opinionated implementation and I'm not sure if we should go the same way.  The TxSNI endpoint requires you to have a directory full of certificates, and assumes you're going to have the same factory backing each one.  A fully flexible implementation would require some way to select which factory to use for each certificate, except that breaks the `connectionMade`-means-TCP-connection nature of our TLS APIs, because by that point you've already been through `buildProtocol`.

Also, `txsni` uses a synchronous mapping, which means certificates need to be loaded synchronously.  Perhaps we would want certificates loaded asynchronously.

So, I don't know.  Should we just roll txsni's endpoint in and then worry about these other issues later?

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_13"></a>|@hynek commented|
|-|-|

I’m starting to be rather positive about external dependencies to Twisted since it lowers the general barriers of contributions.

I would suggest the following:

- move txsni into the twisted organization to make it look more official
- add proper docs at RTFD (which explain the tradeoffs you just mentioned)
- …and a changelog
- give it a proper version that doesn’t look like a pre-alpha (add a development status that reflects that)
- link it from within the TLS docs because no-one will look here

IOW let it look less like a hack and more like an official thing to use and then let’s see what happens.

I would recommend you to just steal my general docs style from e.g. https://github.com/hynek/attrs/tree/master/docs it works very well and you just have to “fill in the blanks”.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_14"></a>|@mithrandi commented|
|-|-|

This is slightly off-topic, but I just ran into an issue implementing SNI in my application that probably needs to be taken into consideration for any SNI implementation that makes its way into Twisted.

The verification settings (VERIFY_PEER, verify callback, etc.) are set on the context, but they are also per-connection state. The connection settings are initialized from the context settings *when the connection is initialized*, meaning that calling `connection.set_context` later does not update them. It is thus necessary to call `connection.set_verify` and possibly `connection.set_verify_depth` at the same time as `connection.set_context`, if you are intending to switch to a context with different verification settings.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Replying to [mithrandi](#note_14):
> This is slightly off-topic, but I just ran into an issue implementing SNI in my application that probably needs to be taken into consideration for any SNI implementation that makes its way into Twisted.
> 
> The verification settings (VERIFY_PEER, verify callback, etc.) are set on the context, but they are also per-connection state. The connection settings are initialized from the context settings *when the connection is initialized*, meaning that calling `connection.set_context` later does not update them. It is thus necessary to call `connection.set_verify` and possibly `connection.set_verify_depth` at the same time as `connection.set_context`, if you are intending to switch to a context with different verification settings.

Oh wow. Uh, hrm. This is an interesting problem :-).  Thank you very much for reporting it.

**Comment by twisted-trac** [comment]
|<img alt="evilalive's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|evilalive commented|
|-|-|

Is there any update on this ticket?

Implementing SNI in twisted would require major work or the implementation is already doable with the current codebase?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Replying to [evilalive](#note_16):
> Is there any update on this ticket?

Nope.  Would you like to contribute something? :)

> Implementing SNI in twisted would require major work or the implementation is already doable with the current codebase?

It would definitely require work, but as txsni shows, you can definitely do it with the current code.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Also of interest: https://txacme.readthedocs.io/en/latest/

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

So, I guess the solution here really ought to be absorbing txsni, but only after txsni implements Deferred support - https://github.com/glyph/txsni/issues/17

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|
